File: algorithms/zig/sorting/topological_sort_of_dag.zig
const std = @import("std");
const HashMap = std.HashMap;
const Vec = std.ArrayList;

pub struct Graph {
    graph: HashMap<usize, Vec<usize>>,
    V: usize,
}

pub fn Graph.init(vertices: usize) Graph {
    return Graph { .graph = HashMap.init(u64, Vec([]usize{})), .V = vertices };
}

pub fn (g: *Graph) addEdge(u: usize, v: usize) void {
    if (!g.graph.containsKey(u)) {
        g.graph.put(u, Vec([]usize{}));
    }
    g.graph.get(u).unwrap().append(v);
}

pub fn (g: *Graph) topologicalSortUtil(v: usize, visited: []bool, stack: *std.ArrayList(usize)) void {
    visited[v] = true;
    if (g.graph.containsKey(v)) {
        const edges = g.graph.get(v).unwrap();
        for (edges) |i| {
            if (!visited[i]) {
                g.topologicalSortUtil(i, visited, stack);
            }
        }
    }
    stack.unshift(v);
}

pub fn (g: *Graph) topologicalSort() void {
    const visited = [_]bool{false} ** g.V;
    var stack = std.ArrayList(usize).init(8);

    for (g.V) |i| {
        if (!visited[i]) {
            g.topologicalSortUtil(i, visited, &stack);
        }
    }

    while (!stack.empty()) {
        const v = stack.pop();
        std.debug.print("{d} ", .{v});
    }
    std.debug.print("\n", .{});
}

pub fn main() void {
    var g = Graph.init(6);
    g.addEdge(5, 2);
    g.addEdge(5, 0);
    g.addEdge(4, 0);
    g.addEdge(4, 1);
    g.addEdge(2, 3);
    g.addEdge(3, 1);
    
    std.debug.print("Following is a Topological Sort of the given graph\n", .{});
    g.topologicalSort();
}

File: algorithms/sorting/topological-sort-of-dag/python/topological_sort_of_dag.py

# recursive topological sorting
# ----------------------------
#Python program to print topological sorting of a DAG
from collections import defaultdict

def main() -> None:
    """Driving code"""
    g= Graph(6)
    g.addEdge(5, 2)
    g.addEdge(5, 0)
    g.addEdge(4, 0)
    g.addEdge(4, 1)
    g.addEdge(2, 3)
    g.addEdge(3, 1)
    
    print ("Following is a Topological Sort of the given graph")
    g.topologicalSort()

    return None

#Class to represent a graph
class Graph:
    """Class to represent a graph."""

    def __init__(self,vertices) -> None:
        self.graph = defaultdict(list) # dictionary containing adjacency List
        self.V = vertices # number of vertices

        return None
 
    # function to add an edge to graph
    def addEdge(self,u,v) -> None:
        """Add an edge to the paragraph."""
        self.graph[u].append(v)

        return None
 
    # A recursive function used by topologicalSort
    def topologicalSortUtil(self,v,visited,stack) -> None:
        """Recursive function sued by topological sort."""
        # Mark the current node as visited.
        visited[v] = True
 
        # Recur for all the vertices adjacent to this vertex
        for i in self.graph[v]:
            if visited[i] == False:
                self.topologicalSortUtil(i,visited,stack)
 
        # Push current vertex to stack which stores result
        stack.insert(0,v)

        return None

    # 
    def topologicalSort(self) -> None:
        """The function to do Topological Sort. It uses
        recursive topologicalSortUtil()."""
        # Mark all the vertices as not visited
        visited = [False]*self.V
        stack =[]
 
        # Call the recursive helper function to store Topological
        # Sort starting from all vertices one by one
        for i in range(self.V):
            if visited[i] == False:
                self.topologicalSortUtil(i,visited,stack)
 
        # Print contents of stack
        print (stack)

        return None

if __name__ == "__main__":
    main()